unignore_outcomes to override ignore_outcome

Bouke opened the pull request:
As discussed in #48, this PR adds a setting ``obey_outcomes``.  For a demo, see [this build](https://travis-ci.org/Bouke/django-user-sessions/jobs/190741996).

> By default, when using ``ignore_outcome`` in your Tox configuration,
any build errors will show as successful on Travis. This might not
be desired, as you might want to control allowed failures inside your
``.travis.yml``. To cater this need, you can set ``obey_outcomes`` to
``False``. This will override ``ignore_outcome`` by setting it to
``False`` for all environments.

Fixes #48

ryanhiebert commented:
Don't worry about the build failure, that's a bug I need to get fixed, that only applies when a PR is opened from another person's repository (and thus can't use my GITHUB_TOKEN).

Bouke commented:
And I see I missed the document in the index that I've added. Actually I added it before, but somehow erased my changes, so had to redo my changes. Luckily I could get most of my work back.

I'm looking into adding some tests for this as well.

ryanhiebert commented:
Sounds good. Sorry for the annoyance with the build failures.

rpkilby commented on src/tox_travis/hooks.py:
```diff
@@ -17,5 +20,17 @@ def tox_addoption(parser):
     """Check for the presence of the added options."""
+
+    if 'TRAVIS' not in os.environ:
+        return
```
There is a similar check in `toxenv.default_toxenv()`. @ryanhiebert - This seems like the correct place to perform the `TRAVIS` env check. Should the duplicate check in `toxenv.default_toxent()` be removed?

Bouke replied:
Yes, I'll remove the additional check.

Bouke commented:
No problem, from the build logs I already suspected an external factor.

I've added a few tests. It was quite some work to get the calculated configuration from tox. I've settled for parsing the output of ``tox --showconfig`` as an INI file after some changes to make it valid INI.

rpkilby commented on docs/obey_outcomes.rst:
```diff
@@ -0,0 +1,30 @@
+.. code-block:: ini
+
+    [travis]
+    obey_outcomes = False
```
I'm seeing what @ryanhiebert is saying here about the name confusion. I would have expected `obey_outcomes` to use a true value, that you're obeying the original outcome and in effect are disabling the `ignore_outcome` setting. Your interpretation here is that the setting is now not obeying the outcome after `ignore_outcome` has been applied. 

Maybe `unignore_outcomes` is the better name? The setting would then expect a true value.

Bouke replied:
Yeah I was even confused when deciding what a `True` value should do here. I'll rename to `unignore_outcomes`.

ryanhiebert commented on docs/unignore_outcomes.rst:
```diff
@@ -0,0 +1,30 @@
+.. code-block:: ini
+
+    [travis]
+    unignore_outcomes = True
```
I see this as a feature of the tox automatic environment discovery, rather than it's own distinct feature. Unless you disagree, let's put this in a new section in the docs for `toxenv` instead of in its own file. As the docs get cleaned up, I forsee things being broken up further for documentation, but let's not try to do that in this PR.

ryanhiebert commented on src/tox_travis/hooks.py:
```diff
@@ -17,5 +20,16 @@ def tox_addoption(parser):
+    travis_reader = tox.config.SectionReader("travis", tox_config)
+    if travis_reader.getbool('unignore_outcomes', False):
+        for env in config.envlist:
+            config.envconfigs[env].ignore_outcome = False
```
Put this into a function in `toxenv.py`, and just call the function from here. This file should be the minimal shims required to interface with Tox hooks.

Bouke commented:
Thanks for the review, I've updated the PR.

ryanhiebert commented:
It looks pretty good to me. I'm going to hold off on merging it until I can get the build failure taken care of. I may ask you to rebase on master after I've done that. I'm not sure I'll get to it today, sorry. Thank you for your work!

ryanhiebert commented:
I'll note that #46 is the bug that needs to get fixed for these tests to pass. #50 is the current PR for fixing it. Thank you @rpkilby for your work on it.

ryanhiebert commented:
Thank you @Bouke and @rpkilby for making this happen today!

Bouke commented:
Nice, great to see it merged! Is it pssible that you push a new release to pypi with this? This was the last blocker for adapting this pacakage. 

> Op 11 jan. 2017 om 05:50 heeft Ryan Hiebert <notifications@github.com> het volgende geschreven:
> 
> Thank you @Bouke and @rpkilby for making this happen today!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ryanhiebert commented:
0.8 is released. https://pypi.python.org/pypi/tox-travis
